### PR TITLE
feat:typescript migration

### PR DESCRIPTION
**Pull request by cjpillsbury**
baseline for the migration efforts

### CODE REVIEW

File: src/js/constants.ts [hunk 1]
```diff
@@ -0,0 +1,242 @@
+export const MediaUIEvents = {
```
Comment (cjpillsbury) [inline src/js/constants.ts:1]: tried to keep these as tight as possible to make derivation of types automatic and easier to maintain as we add things.

File: src/js/utils/server-safe-globals.ts [hunk 2]
```diff
@@ -79,59 +84,18 @@ const globalThisShim = {
...
- * exitPictureInPicture?,
- * pictureInPictureEnabled?,
- * requestPictureInPicture?,
- * addEventListener?,
- * removeEventListener?,
- * } }
- */
-export const Document =
-  isServer && !isShimmed ? documentShim : globalThis.document;
+export const GlobalThis: typeof globalThis =
```
Comment (cjpillsbury) [inline src/js/utils/server-safe-globals.ts:94]: updated these to "lie" on their export types as if they are always running in the browser and not minimal shim impls when on e.g. node. That does mean we won't get type errors for cases where we have failed to shim/polyfill.

File: src/js/media-chrome-button.ts [hunk 3]
```diff
@@ -95,19 +95,25 @@ template.innerHTML = /*html*/ `
 class MediaChromeButton extends globalThis.HTMLElement {
```
Comment (cjpillsbury) [inline src/js/media-chrome-button.ts:95]: I migrated this one component as part of proving out/validating any issues.

File: src/js/experimental/media-chrome-option.js [hunk 4]
```diff
@@ -109,7 +109,7 @@ class MediaChromeOption extends globalThis.HTMLElement {
 
   enable() {
     if (!this.hasAttribute('tabindex')) {
-      this.setAttribute('tabindex', -1);
+      this.tabIndex = -1;
```
Comment (heff) [inline src/js/experimental/media-chrome-option.js:112]: This is safe, right?

File: src/js/media-captions-button.js [hunk 5]
```diff
@@ -155,4 +155,5 @@ if (!globalThis.customElements.get('media-captions-button')) {
   );
 }
 
+export { MediaCaptionsButton };
```
Comment (heff) [inline src/js/media-captions-button.js:158]: Why is this needed?

**Review by heff — APPROVED** [review]
A few questions but looks good to me. There's some dark magic in the contants types I don't fully grok if it's worth getting another set of eyes on.

File: src/js/experimental/media-chrome-option.js (hunk 4 above)
Comment (cjpillsbury) [reply src/js/experimental/media-chrome-option.js:112]: should be, but I'll double-confirm 😅

Comment (cjpillsbury) [reply src/js/experimental/media-chrome-option.js:112]: yup confirmed!

File: src/js/media-captions-button.js (hunk 5 above)
Comment (cjpillsbury) [reply src/js/media-captions-button.js:158]: Plausibly not needed if we migrate to full TypeScript. The `@typedef` imports were getting confused from using the default module export. This is also a pattern in some other of our component modules. There are good reasons to do named exports/both (tl;dr - different build tools can get confused). Might be a good candidate for our development principles (in whatever direction we decide).

File: src/js/media-chrome-button.ts (hunk 3 above)
Comment (littlespex) [reply src/js/media-chrome-button.ts:95]: The issue is that when the utility functions that except `HTMLElement`s are converted to TS, they still don't consider these true `HTMLElement`s:

Screenshot 2024-06-17 at 16 56 34

Comment (littlespex) [reply src/js/media-chrome-button.ts:95]: Do we know of a specific environment/configuration that breaks without `server-safe-globals.js`? I have a few ideas for workarounds, but I have no way to verify them without a test case.

Comment (luwes) [reply src/js/media-chrome-button.ts:95]: yep, a Next.js app should reproduce the issues that come up without patching the globals.
an example here https://github.com/muxinc/media-chrome/tree/main/examples/nextjs-with-typescript

Comment (littlespex) [reply src/js/media-chrome-button.ts:95]: The next's doesn't need the globals at all:

https://github.com/littlespex/media-chrome/pull/4

**Comment by littlespex** [comment]
@cjpillsbury I got permissions errors when trying to commit to this branch, or create a new one, so I copied all of this over to my fork and used it as a base for https://github.com/muxinc/media-chrome/pull/931.

**Comment by luwes** [comment]
closed by https://github.com/muxinc/media-chrome/pull/931
